distributed/shuffle/_arrow.py

```diff
@@ -6,62 +6,64 @@
     import pyarrow as pa
 
 
-def dump_batch(batch: pa.Buffer, file: BinaryIO, schema: pa.Schema) -> None:
+def dump_shards(shards: list[pa.Table], file: BinaryIO) -> None:
     """
-    Dump a batch to file, if we're the first, also write the schema
+    Write multiple shard tables to the file
 
-    This function is with respect to the open file object
+    Note: This function appends to the file and dumps each table as an individual stream.
+    This results in multiple end-of-stream signals in the file.
 
     See Also
     --------
-    load_arrow
+    load_partition
     """
-    if file.tell() == 0:
-        file.write(schema.serialize())
-    file.write(batch)
+    import pyarrow as pa
+
+    for table in shards:
+        with pa.ipc.new_stream(file, table.schema) as writer:
+            writer.write_table(table)
 
 
-def load_arrow(file: BinaryIO) -> pa.Table:
-    """Load batched data written to file back out into a table again
+def load_partition(file: BinaryIO) -> pa.Table:
+    """Load partition data written to file back out into a single table
 
     Example
     -------
-    >>> t = pa.Table.from_pandas(df)  # doctest: +SKIP
+    >>> tables = [pa.Table.from_pandas(df), pa.Table.from_pandas(df2)]  # doctest: +SKIP
     >>> with open("myfile", mode="wb") as f:  # doctest: +SKIP
-    ...     for batch in t.to_batches():  # doctest: +SKIP
-    ...         dump_batch(batch, f, schema=t.schema)  # doctest: +SKIP
+    ...     for table in tables:  # doctest: +SKIP
+    ...         dump_shards(tables, f)  # doctest: +SKIP
 
     >>> with open("myfile", mode="rb") as f:  # doctest: +SKIP
-    ...     t = load_arrow(f)  # doctest: +SKIP
+    ...     t = load_partition(f)  # doctest: +SKIP
 
     See Also
     --------
-    dump_batch
+    dump_shards
     """
     import pyarrow as pa
 
+    shards = []
     try:
-        sr = pa.RecordBatchStreamReader(file)
-        return sr.read_all()
-    except Exception:
-        raise EOFError
-
-
-def list_of_buffers_to_table(data: list[bytes], schema: pa.Schema) -> pa.Table:
+        while True:
+            sr = pa.RecordBatchStreamReader(file)
+            shards.append(sr.read_all())
+    # Since we write multiple streams to the same file, we have to read until
+    # there is nothing to read anymore. At that point, pa.ArrowInvalid is raised
+    except pa.ArrowInvalid as e:
+        if str(e) == "Tried reading schema message, was null or length 0":
+            return pa.concat_tables(shards)
```

Comment:
Parsing this error message seems brittle. Is there some other check we can do on `file` to make sure we're read it all?

Reply:
I forgot that we can seek the end of the open file object, I'm using that instead now.